In [1]:
import math
import time

def naiveSum(n):
    start_time=time.time()
    s=0
    for i in range(0,n):
        s += math.pi
    end_time=time.time()
    totaltime=end_time-start_time
    return s, totaltime

def compensatedSum(n):
    start_time=time.time()
    s=0
    e=0
    for i in range(0,n):
        temp=s
        y=math.pi+e
        s=temp+y
        e=(temp-s)+y
    s += e
    end_time=time.time()
    totaltime=end_time - start_time
    return s, totaltime

def main():
    N=[10**i for i in range(6,10)]

    naivesums=[]
    naivesumtimes=[]
    compensatedsums=[]
    compensatedsumtimes=[]
    exactsums=[]
    
    for n in N:
        naivesum, naivesumtime = naiveSum(n)
        naivesums.append(naivesum)
        naivesumtimes.append(naivesumtime)

        compensatedsum, compensatedsumtime = compensatedSum(n)
        compensatedsums.append(compensatedsum)
        compensatedsumtimes.append(compensatedsumtime)

        exactsums.append(n * math.pi)
    
    for i in range(0,4):
        print(f'N={N[i]} \n\nNaive Summation \nSum={naivesums[i]} \nExact Sum={exactsums[i]} \nTime={naivesumtimes[i]} \n\nCompensated Summation \nSum={compensatedsums[i]} \nExact Sum={exactsums[i]} \nTime={compensatedsumtimes[i]}\n\n')
        
main()

N=1000000 

Naive Summation 
Sum=3141592.6535849427 
Exact Sum=3141592.653589793 
Time=0.18497657775878906 

Compensated Summation 
Sum=3141592.653589793 
Exact Sum=3141592.653589793 
Time=0.12549924850463867


N=10000000 

Naive Summation 
Sum=31415926.543478474 
Exact Sum=31415926.535897933 
Time=0.7682390213012695 

Compensated Summation 
Sum=31415926.535897933 
Exact Sum=31415926.535897933 
Time=1.358506679534912


N=100000000 

Naive Summation 
Sum=314159265.624272 
Exact Sum=314159265.3589793 
Time=7.613440036773682 

Compensated Summation 
Sum=314159265.3589793 
Exact Sum=314159265.3589793 
Time=12.411301851272583


N=1000000000 

Naive Summation 
Sum=3141592632.494384 
Exact Sum=3141592653.589793 
Time=67.89723825454712 

Compensated Summation 
Sum=3141592653.589793 
Exact Sum=3141592653.589793 
Time=118.90951013565063




|                  | 10^6                | 10^7               | 10^8               | 10^9               |
|------------------|---------------------|--------------------|--------------------|--------------------|
| Naive Sum        | 3141592.6535849427  | 31415926.543478474 | 314159265.624272   | 3141592632.494384  |
| Naive Time       | 0.06725859642028809 | 0.6430079936981201 | 6.44527530670166   | 64.46708989143372  |
| Compensated Sum  | 3141592.653589793   | 31415926.535897933 | 314159265.3589793  | 3141592653.589793  |
| Compensated Time | 0.11365270614624023 | 1.1371495723724365 | 11.355765104293823 | 113.49001717567444 |
| Exact Value      | 3141592.653589793   | 31415926.535897933 | 314159265.3589793  | 3141592653.589793  |

# Absolute and Relative Error of S<sub>n</sub>=10<sup>6</sup>

## Naive Sum
- E<sub>a</sub> = 3141592.653589793 - 3141592.653584942 = 0.000004851
- E<sub>r</sub> = (0.000004851 / 3141592.653584942) * 100 = 1.5441212578799529752952934941577928476713789717719359 * 10<sup>-10</sup>

## Compensated Sum
- E<sub>a</sub> = 3141592.653589793 - 3141592.653589793 = 0
- E<sub>r</sub> = (0 / 3141592.653589793) * 100 = 0

# Absolute and Relative Error of S<sub>n</sub>=10<sup>7</sup>

## Naive Sum
- E<sub>a</sub> = 31415926.535897933 - 31415926.543478474 = 0.007580541
- E<sub>r</sub> = (0.007580541 / 31415926.543478474) * 100 = 2.4129611423393205261601803571089421318874089023358493 * 10<sup>-8</sup>

## Compensated Sum
- E<sub>a</sub> = 31415926.535897933 - 31415926.535897933 = 0
- E<sub>r</sub> = (0 / 31415926.535897933) * 100 = 0

# Absolute and Relative Error of S<sub>n</sub>=10<sup>8</sup>

## Naive Sum
- E<sub>a</sub> = 314159265.3589793 - 314159265.624272 = 0.2652927
- E<sub>r</sub> = (0.2652927 / 314159265.624272) = 8.4445289071080461373687190849919864670357012206871846 * 10<sup>-8</sup>

## Compensated Sum
- E<sub>a</sub> = 314159265.3589793 - 314159265.3589793 = 0
- E<sub>r</sub> = (0 / 314159265.3589793) * 100 = 0

# Absolute and Relative Error of S<sub>n</sub>=10<sup>10</sup>

## Naive Sum
- E<sub>a</sub> = 3141592653.589793 - 3141592632.494384 = 21.095409
- E<sub>r</sub> = (21.095409 / 3141592632.494384) = 6.7148772828800905175348683876016429449863661872396001 * 10<sup>-7</sup>

## Compensated Sum
- E<sub>a</sub> = 3141592653.589793 - 3141592653.589793 = 0
- E<sub>r</sub> = (0 / 3141592653.589793) * 100 = 0

# Analysis/Discussion
After running the test, the most prominent observation is that the error for the compenstated summation is zero while the error for the naive summation is non-zero. Upon further inspection, it can be seen that both the absolute and realtive errors for the naive summation is increasing at an exponentailly rate. However, the calculation time for the compensated summation is consistently almost double the calculation time for the naive summation and this becomes more evident as the number of terms being added increases. Overall, these errors match up with the expected errors for each of the two algorithms.